### ISSUE TITLE
"Cannot load file from outside of service folder" deprecation warning

### ISSUE THREAD

**Issue by m-radzikowski** [serverless/serverless#9095]
In the official docs for [referencing variables in other files](https://www.serverless.com/framework/docs/providers/aws/guide/variables#reference-variables-in-other-files) as an example there is a reading file from a parent directory. In the mono-repo approach with multiple stacks, I believe this is a fairly often approach.

But from one of the latest releases I'm getting this warning:

```
Serverless: Deprecation warning: Variables resolver reports following resolution errors:
              - Cannot resolve variable at "custom.project": Cannot load file from outside of service folder
            From a next major it we will be communicated with a thrown error.
            Set "variablesResolutionMode: 20210219" in your service config, to adapt to this behavior now
            More Info: https://www.serverless.com/framework/docs/deprecations/#NEW_VARIABLES_RESOLVER
```

Changing to reading files only from the same directory would require using some extra plugins to maintain the current behavior. And I'm sure I'm not the only one that shares some common config between the stacks.

<details>
<summary><code>serverless.yml</code></summary>

```yaml
service: test

custom:
  project: ${file(../config.yml):project}
```

</details>

<b>Installed version</b>

```
Framework Core: 2.29.0 (local)
Plugin: 4.5.0
SDK: 4.2.0
Components: 3.7.2
```

**Comment by chuckthepiranha** [comment]
I am having a similar problem I think except mine is that I can't include a vtl file as a mapping template. For example: 
`request:
       template:
            application/x-www-form-urlencoded: ${file(velocity-language-templates/x-www-form-urlencoded.vtl)}`
gives me this warning: 
`- Cannot resolve variable at "custom.apiGatewayServiceProxies.0.http.request.template.application/x-www-form-urlencoded": Cannot parse "x-www-form-urlencoded.vtl": Unsupported file extension`

**Comment by Borduhh** [comment]
Just to speak for the people that have a mono-repo setup. This file structure is used A LOT. I personally like the structure of having each service be essentially its own contained unit so that when we are updating services, the code is all in one spot for the most part (excluding libs, etc.).

```
serverless.common.yml
|_ services
    |_ service-a
        |_ serverless.yml
    |_ service-b
        |_ serverless.yml
    |_ service-c
        |_ severless.yml
```

> Have you considered putting all service configurations at root of the project as e.g. `service-a.yml`, `service-b.yml` etc. and deploying it with help of `--config` option ?

Can you explain a little more about what that entails?

**Comment by m-radzikowski** [comment]
@medikoo thanks for the explanation of the reasoning behind this change.

> we expect the project configuration to not reach beyond the project folder, as that's an anti-pattern for security and project portability reasons.

Agree that reaching outside the project folder would be a terrible idea, and this prevents it. But in monorepo the service root is not the project root.

> Note also that all paths in configuration are always resolved against root of the service

Yes, and I must admit that this creates quite confusing paths when you need to self-referenece something in the "common" config. If we could solve this somehow along the way, it would be great.

> Have you considered putting all service configurations at root of the project as e.g. `service-a.yml`, `service-b.yml` etc. and deploying it with help of `--config` option ?

If I understand correctly, it would be changing the current structure (like posted by @Borduhh) to something like this:

```
serverless.common.yml
serverless.a.yml
serverless.b.yml
serverless.c.yml
|_ services
    |_ service-a
        |_ ...
    |_ service-b
        |_ ...
    |_ service-c
        |_ ...
```

I see few problems here.

Firstly, bloated root project dir.

Secondly, paths in each serverless file to its resources would change from simple `$file(smth.yml)` to `$file(services/service-a/smth.yml)`. It's not only longer and less readable, but also would open an easy way to "reuse" things from other services. If the path is long anyway, someone could import some resources from "service-a" in "service-b". Yes, it is possible right now (in v2), but since serverless file is in its service dir, it's much less "natural" to do something like this. I hope I explained what I mean by this.

And lastly, with monorepo you can use lerna to deploy all services. If each service directory has `serverless.yml` and `package.json`, you can use lerna from the root dir to execute the deployment in all services. Each of the service package.json files has simple script `"deploy": "sls deploy"`. After that change, it would have to be `"deploy": "sls deploy --config ../../serverless.a.yml`. Quite terrible to be honest.

Also, when you google "serverless monorepo", you got [all](https://serverless-stack.com/chapters/organizing-serverless-projects.html) [those](https://github.com/erezrokah/serverless-monorepo-app) [examples](https://medium.com/serverlessguru/how-do-i-structure-a-monorepo-serverless-project-with-the-serverless-framework-79b22de0abb6) from the first results. I'm 100% into improvements and breaking changes if they are needed, but would be nice to make sure that it actually improves a project, not only makes a lot of articles outdated 😉 

**Proposed solution**

Some solution to keep advantages of local-only path resolution and have common config at the same time would be to provide the ability to extend a serverless file. I think that was proposed several times, like in #8531. Then the structure would stay the same:

```
serverless.common.yml
|_ services
    |_ service-a
        |_ serverless.yml
    |_ service-b
        |_ serverless.yml
    |_ service-c
        |_ severless.yml
```

And each `serverless.yml` would have something like:

```
service: a
extends: ../../serverless.common.yml
```

This is similar to what exists in other tools, like ESLint, tsconfig, etc.

With `extends`, we could define some `custom` and `provider` fields that are the same in all services. It would simplify my stacks a lot, and I think others would agree.

It would be brilliant if paths in `serverless.common.yml` would be resolved against its location, not the `service-a` dir 😉 

I know this is not a trivial thing to add. But in my opinion, it would make adapting to breaking changes in v3 totally worth it! Yes, I would have to adapt anyway. But I would be much happier to do so!

To simplify the implementation, maybe the `extends` could be limited only to `custom` and `provider` fields. Or just to `custom`.

What do you think?

**Comment by medikoo** [comment]
>  But in monorepo the service root is not the project root.

True, I think we can solve it, I've proposed something at the bottom of this comment.

> Yes, and I must admit that this creates quite confusing paths when you need to self-referenece something in the "common" config. If we could solve this somehow along the way, it would be great.

This problem was [once raised](https://github.com/serverless/serverless/issues/3233) and as long as we have one service configuration, and not multiple independent configurations which cross reference each other, then it's hard to address nicely.

With `self:` it's always main service configuration that's addressed, and not configuration part that's maintained in given file (technically in `self` resolution there's no file system concept involved)

There was a once proposal to introduce `local:` to address this use case https://github.com/serverless/serverless/issues/3233, still it'll imply a need to resolve given configuration file on its own, which is very different from how file imports are handled (currently we just copy bits referenced by `file` onto main configuration, and after doing that any variables found in those bits are resolved in context of main configuration - at this point context of file from which given configuration bits were imported is lost)

Here some DX problem is that `self:` is not intuitive name for what it does. It'll probably work better as `config:`.

Anyway I'll give it a second though, maybe with new variables resolver (we have now in place) there's a room to somehow achieve capability of `local:`. It requires a deeper analysis.

> After that change, it would have to be "deploy": "sls deploy --config ../../serverless.a.yml. Quite terrible to be honest.

Not exactly, with this change, deployments will have to be done from project root folder as `sls deploy --config services/service-a/serverless.a.yml` (but then yes all paths in `serlveress.a.yml` will have to also be based against root of a project)

If sub projects depends on resources of a root of the project, then I believe they also should be operated from root of the project (working directory marks the boundary in which given command should work).

I believe with Lerna typical case, is that commands issued on subprojects do not reach out of its folders (at least not by direct fs traversal, e.g. npm packages maintained that way cross-reference each other by their names,  and through configuration in `package. json` and not via paths prefixed with `../`).

--- 

Anyway as I think of it. I think we can improve the experience, by internally recognizing two dir contexts _project dir_ (project boundary), and _base dir_ (path against which fs paths should be resolved). Currently in Framework there's just _service dir_ (which is assumed as both _project dir_ and _base dir_)

It'll probably be not that difficult to introduce to Framework _project dir_ concept. Project dir will be the one from which command is invoked, and Base dir will be the on in which service configuration is placed (internally it'll stay as _service dir_).

Having that you will have to deploy service as `sls deploy --config services/service-a/serverless.yml`, but all paths in `services/service-a/serverless.yml` will be resolved against `services/service-a` folder. It's just this directory that will be by default packaged for lambdas. But you'll be allowed to reach out outside up to root project folder for file path imports (with note that it won't work for packaging - still we probably can fix it with additional improvement)

@m-radzikowski @pgrzesik what do you think?

**Comment by pgrzesik** [comment]
Thanks @medikoo for proposal - I have one thing to add related to:

> It's just this directory that will be by default packaged for lambdas. But you'll be allowed to reach out outside up to root project folder for file path imports (with note that it won't work for packaging - still we probably can fix it with additional improvement)

I can imagine that in such monorepo setups, there will be parts of code that will live outside of the _service dir_ that are shared across multiple services and if we'd like to support it "properly", then packaging in such setups also needs to work reliably. 

@m-radzikowski is that the case in your situation? Do you have some shared libraries/modules that live outside of individual service directories?

**Comment by medikoo** [comment]
> I can imagine that in such monorepo setups, there will be parts of code that will live outside of the service dir that are shared across multiple services and if we'd like to support it "properly", then packaging in such setups also needs to work reliably.

Definitely, still that part also doesn't work now. I assume that either @m-radzikowski uses some custom packaging mechanims (coming from plugin) which allows to include content from top folder, or just takes what's in service folder (as that what Framework does currently, and afaik that cannot be really tweaked without overrriding packaging logic completely)

**Comment by m-radzikowski** [comment]
> I can imagine that in such monorepo setups, there will be parts of code that will live outside of the service dir that are shared across multiple services and if we'd like to support it "properly", then packaging in such setups also needs to work reliably.

> Definitely, still that part also doesn't work now. I assume that either @m-radzikowski uses some custom packaging mechanims (coming from plugin) which allows to include content from top folder

Yes, with plugins like `serverless-webpack` or `serverless-esbuild` we are able to load some common code from shared dir.

My current structure with lerna looks like this:

```
package.json
lerna.json
|_ services
    |_ common
        |_ src
            |_ logger.ts
    |_ service-a
        |_ package.json
        |_ serverless.yml
        |_ src
            |_ ...
    |_ service-b
        |_ package.json
        |_ serverless.yml
        |_ src
            |_ ...
```

Now, with those plugins, I can reference the `logger` from the common dir in code of both services.

> Having that you will have to deploy service as `sls deploy --config services/service-a/serverless.yml`, but all paths in `services/service-a/serverless.yml` will be resolved against `services/service-a` folder.

One issue I have with this solution is that the deployment management of all services would be done from the root dir, not from the individual Lerna packages (our individual service packages) itself.

Now in root `package.json` the `deploy` script is just calling Lerna to run the `deploy` script in all services, so each service is responsible for running own deployment command. It's just `sls deploy` in most cases, but sometimes needs additional actions. Configuring the deployment command on the service level instead of the root project level makes the most sense for me.

I totally understand and agree with the scope access argument. At the same time, I would like to find an approach that will result in an improved workflow, and not degrade it. And I have no problem with changing my workflow (or, better to say, workflows in my company) 😉 

What about, instead of running the command from the project level, passing an argument like `--project-root`, that will define the scope? I'm just throwing ideas here 😄

**Comment by medikoo** [comment]
Thanks @m-radzikowski for explanations

> Now, with those plugins, I can reference the logger from the common dir in code of both services.

Just wondering, where do you have `serverless-webpack` or `serverless-esbuild` installed, is this in top level project folder, on they're installed for each service individually? Also where do you host the bundler configuration that states that `logger` should be included with lambda bundle, is this in service folder or project folder?

**Comment by m-radzikowski** [comment]
> > Now, with those plugins, I can reference the logger from the common dir in code of both services.
> 
> Just wondering, where do you have `serverless-webpack` or `serverless-esbuild` installed, is this in top level project folder, on they're installed for each service individually? Also where do you host the bundler configuration that states that `logger` should be included with lambda bundle, is this in service folder or project folder?

Plugins that are used in multiple packages - in `services/common/package.json`. I guess they could be in root level `package.json` as well, but that allows to keep separation between project dependencies (things like lerna, typescript, eslint, husky) and services dependencies.

I'm using yarn, so most dependencies are installed into top-level `node_modules`, no matter in which `package.json` they are declared (if there are no dependency version conflicts).

With yarn + webpack/esbuild no additional config is needed to bundle dependencies from `common` together with a particular Lambda code. Having `services/common/package.json`:

```
{
  "name": "@myproject/common",
  ...
}
```

in another service I import the logger just as:

```typescript
import {logger} from '@myproject/common/src/logger';
```

And that's it, it will be bundled with code.

**Comment by medikoo** [comment]
Thanks @m-radzikowski for clarification. So it's as I thought, that you do not reach out from service root by traversing paths (e.g `../../services/common..`) but you're relying on built-in lerna intelligence which allows to work address package dependencies as if they're installed in service `node_modules`.

Problem we have here, that such mono-repo handling requires special tooling (as Lerna), as Node.js doesn't support it out of a box.

For lambda packaging you've solved cross-package links via `serverless-webpack` or `serverless-esbuild` plugins which I believe have built in support for Lerna (or it's Lerna that has built-in support for them), but you're missing something that will add similar intelligence to serverless service config resolution.

I think that ideally, as this is a custom environment, this could solved externally by something as `serverless-lerna` plugin (which should provide support for such setup in full capacity, so every aspect of a Framework work with it out of a box, also packaging). Having that you wouldn't reach for other config part via `${file(../../services/common/config.yml):project}` but rather via `${file(@myproject/common/src/config.yml)}`. What do you think? Will that work for you?

**Comment by m-radzikowski** [comment]
Yes, it would be the perfect solution. Even for projects having multiple stacks and not using Lerna currently, migrating to using Lerna to achieve common config files would be probably beneficial.

Since you suggest making this as a separate plugin, which makes sense as it's not a built-in Node feature, I guess I would have to develop it to allow for seamless migration to serverless v3? 😉  Is there a planned roadmap for v3, with an estimated release date?

**Comment by medikoo** [comment]
> Is there a planned roadmap for v3, with an estimated release date?

There's no date set in stone, we hope to release it sometime before mid of year.

On our side, we will do our best to not go into v3 without providing a final solution for this use case.

In worst case scenario we will postpone deprecation, temporarily relax path restriction, and then re-introduce deprecation and work it out on v3 grounds.

**Comment by ianpogi5** [comment]
> On our side, we will do our best to not go into v3 without providing a final solution for this use case.
> 
> In worst case scenario we will postpone deprecation, temporarily relax path restriction, and then re-introduce deprecation and work it out on v3 grounds.

Thank you @medikoo I can breath much easier now.

**Comment by medikoo** [comment]
I've updated main description with a solution proposal, that outlines all steps needed, to I believe a really good multi-service repo support. Note that just first two (small) steps will fix the source of this issue.

@pgrzesik @m-radzikowski what do you think about that?

**Comment by pgrzesik** [comment]
Thanks for a detailed outline @medikoo :bow:. I have one question related to the proposal, when `projectDir` is involved, all `file(..)` references in `serverless.yml` will be resolved relatively to that `projectDir` or relatively to `serviceDir` ?

**Comment by medikoo** [comment]
> Thanks for a detailed outline @medikoo 🙇. I have one question related to the proposal, when projectDir is involved, all file(..) references in serverless.yml will be resolved relatively to that projectDir or relatively to serviceDir ?

It'll be resolved same as it happens now, so relatively to `serviceDir` (folder in which `serverless.yml` is placed),  It's just that you'd be able to pick file from parent folder (assuming it's in a boundary of `projectDir`)

**Comment by pgrzesik** [comment]
Thanks for the clarification as I wasn't 100% sure.

**Comment by m-radzikowski** [comment]
Your proposal looks great, also all the related features you listed seem like a good direction. I have a question though - I don't fully understand the relationship between the `baseDir` and `sourceDir`? The handler function must be in the sources anyway, right? Maybe if the handler path would be relative to source dir, there could be only one param, which would reduce introduced complexity? What is the use case for setting both of those 2 params?

**Comment by medikoo** [comment]
> I don't fully understand the relationship between the `baseDir` and `sourceDir`? 

The default packaging approach in a Framework it so simply package a `serviceDir` (with automatic exclusion of obvious parts as _dev_ dependencies, external plugins and service configuration).

With `baseDir` you'd be able to change, which folder should be treated as base for packaging. So e.g. in case of multi-repo, you may point the parent folder of the `serviceDir` (`serviceDir` points the folder in which `serverless.yml` is placed).

`baseDir` simply states what is the _root_ folder. So folder which by default is packaged as whole (with some obvious parts excluded), and against which all package related paths (as `handler` or `package` settings) need to be configured.

Now with `sourceDir` you may state that you just want some specific folder within `baseDir` to be packaged by _default_, but that doesn't switch the _root_ folder. So still all the paths in configuration need to be relative against `basePath`, and in lambda it's `basePath` that will reflect _root_.

For example in following structure:

```
projectRoot/
  service1/
    lambda1/
      handler.js
    lambda2/
    serverless.yml
  service2/
  common/
```

For _service1_ by default `projectDir`, `serviceDir`, `baseDir` and `sourceDir` is `projectRoot/service1` folder. Which means that just this folder is packaged, and you have no means (even with `pattern.include`) to include any content from parent folder (as that reaches out beyond `baseDir`).

After setting `projectDir` and `baseDir` to `..` by default all `projectRoot` folder will be packaged. So now you have access to common parts, but also you're packaging whole project for each lambda. You may apply exclude rules to exclude other services, but that may not be convenient, you may just want to _include_ `service1` and `common` instead.

Currently developers usually solve it by using `!**` pattern to exclude everything and then just including what is expected to be packaged. `sourceDir` will be a dedicated setting to achieve same. e.g. by setting `sourceDir: service1/lambda1` and adding `patterns: common/**` end content of the lambda will be as:

```
service1/
  lambda1/
    handler.js
common/
```

**Comment by Bolik777** [comment]
I have maybe provocative question, but why we need to over-complicate the setup so much and not leave the variable resolving as it is? Does it improve the security so much? I don't also understand the portability argument...

We use the monorepo a lot, because:
1. it allows us to keep all the dependencies in a single place (root package.json)
2. have cross service resources in a root folder. Here i mean stuff like configs, tags, custom variables etc. -> DRY. In big projects with tons of services it is error prone to repeat the "shared" variables in every single service.
3. have all service dependencies in a single folder, it makes the folder/project structure clean. 
we don't use Lerna, just pure serverless + serverless bundle plugin.
the overall goal of the serverless framework is to simplify the deployments in any cloud and provide enough flexibility for projects of any size.

**Comment by medikoo** [comment]
> We use the monorepo a lot, because:

We understand that use case, and we do not question it anywhere

> I don't also understand the portability argument...

If portability is not ensured it's way more likely to create a service that's cumbersome to setup in other environments.

Anyway, that's a minor here, and focusing just on that argument is bike-shedding. Note, it's not just about resolution of `file` variable sources. This proposal ensures that all Framework functionalities (packaging etc.) can work with monorepo structure without significant fiddling and idea as outlined in main description is probably best way to achieve that.

**Comment by alexfpoole** [comment]
I'm just doing a `ln -s ../base-service/ base-service` in each service to get around this

**Comment by tux86** [comment]
It's really blocking for my monorepo project.  Today I have 5 micorservices and all common plugins configs are shared in serverless-common.yml.


Abstract monorepo project structure: 

`services/
  - service-a/
     - serverless.yaml
  - service-b/
     - serverless.yaml
  - ..../
    - serverless.yaml
    
 serverless-common.yml <-- common variables /  plugins config 
 `

The problem that i don't have only variables to share with my micro-services (I could go with .env files with no problems) but I have plugins configuration like shown below.

```
bundle:
  sourcemaps: true
  caching: true
  linting: false
  packager: yarn
  forceExclude:
    - aws-sdk
    - chrome-aws-lambda

deploymentBucket:
  name: serverless-deploys-${self:provider.stage}-mycompanydomain
  blockPublicAccess: true

```

This kind of config could not be done with .env.

I couldn't go to version 3 without this feature.

I suggest to not reinvent the wheel to solve this by adding an "extend:" entry in serverless.yaml like typescript configuration tsconfig.json. This will be very helpful for monorepo projects.

**Comment by medikoo** [comment]
@tux86 we've already proposed a solution (it's outlined in description), and I believe it addresses your use case well. Have you read it?

**Comment by tux86** [comment]
> @tux86 we've already proposed a solution (it's outlined in description), and I believe it addresses your use case well. Have you read it?

 I think i don't . I will check that. Thank you.

**Comment by OndeVai** [comment]
Any update on this issue? It seems draconian to not allow file paths from levels above. If someone includes a file path that's incorrect, they will find out quickly enough. I don't agree that the framework should be checking this.

**Comment by slootjes** [comment]
Another mono-repo user here that would be affected by this. Please revert this.

**Comment by ivh-mm** [comment]
Also +1 to revert this. Or at least provide an option to opt out of this requirement.

**Comment by poldark101** [comment]
Thanks for all of the info on this thread. 
@medikoo, hope you can help
Following your solution, what file in the top dir/root folder is the config property 'projectDir' defined in, can you give an example?  Does this then override the projectDir in the Serverless code?  I am also using a mono repo.  Serverless.yml files include resources that reference files outside of the services directory for example in the 'deploy' folder seen below.

deploy/   (components and scripts for deployment)
node_modules/
services/
  ---api/
     -serverless.yml
  ---infra/
     -serverless.yml
package.json

**Comment by jlpetracca** [comment]
revert this please

**Comment by excenter** [comment]
If I write a common yaml file to contain shared resources, and put that in a static path in relationship to several `serverless.yml`s within a monorepo, and it pulls the configs from the file path I told it to look in, that is much less confusing than the weird workarounds I would have to engineer to avoid this error that shouldn't be an error.

**Comment by excenter** [comment]
> > I don't fully understand the relationship between the `baseDir` and `sourceDir`?
> 
> The default packaging approach in a Framework it so simply package a `serviceDir` (with automatic exclusion of obvious parts as _dev_ dependencies, external plugins and service configuration).
> 
> With `baseDir` you'd be able to change, which folder should be treated as base for packaging. So e.g. in case of multi-repo, you may point the parent folder of the `serviceDir` (`serviceDir` points the folder in which `serverless.yml` is placed).
> 
> `baseDir` simply states what is the _root_ folder. So folder which by default is packaged as whole (with some obvious parts excluded), and against which all package related paths (as `handler` or `package` settings) need to be configured.
> 
> Now with `sourceDir` you may state that you just want some specific folder within `baseDir` to be packaged by _default_, but that doesn't switch the _root_ folder. So still all the paths in configuration need to be relative against `basePath`, and in lambda it's `basePath` that will reflect _root_.
> 
> For example in following structure:
> 
> ```
> projectRoot/
>   service1/
>     lambda1/
>       handler.js
>     lambda2/
>     serverless.yml
>   service2/
>   common/
> ```
> 
> For _service1_ by default `projectDir`, `serviceDir`, `baseDir` and `sourceDir` is `projectRoot/service1` folder. Which means that just this folder is packaged, and you have no means (even with `pattern.include`) to include any content from parent folder (as that reaches out beyond `baseDir`).
> 
> After setting `projectDir` and `baseDir` to `..` by default all `projectRoot` folder will be packaged. So now you have access to common parts, but also you're packaging whole project for each lambda. You may apply exclude rules to exclude other services, but that may not be convenient, you may just want to _include_ `service1` and `common` instead.
> 
> Currently developers usually solve it by using `!**` pattern to exclude everything and then just including what is expected to be packaged. `sourceDir` will be a dedicated setting to achieve same. e.g. by setting `sourceDir: service1/lambda1` and adding `patterns: common/**` end content of the lambda will be as:
> 
> ```
> service1/
>   lambda1/
>     handler.js
> common/
> ```

If this is 
> @tux86 we've already proposed a solution (it's outlined in description), and I believe it addresses your use case well. Have you read it?

That does not meet my use case. It's overly complicated compared to "there are serveral serverless.yml files that reference these shared configs" 
This change seems like a mistake that causes pain for wins that can be achieved with logging or a different approach.

**Comment by medikoo** [comment]
> If this is

Proposed solution is in main issue description.

_To all commenters_:

Serverless Framework, never (at any stage) supported multi-service repositories. What you have currently is:
1. Some loose parts which you can swing in any direction (as they work without any security checks and validation implied)
2. Core logic (like packaging) that simply do not work with multi-service repos (you cannot package files outside of service dir) which you patch by using plugins which override what we have in core.

Goal of this issue is to make multi-service repo support a first class citizen in a Framework, so it works well, without any special overhead and without a need of using any plugins (again please see top issue description for full proposal).

Those which are specifically bothered by deprecation message that is shown when in variables you reach for files outside of service directory - this will be addressed with first PR (after it's in, you'll make deprecation message gone simply by configuring `projectDir` property in service config, and nothing more will have to be done)

**Comment by medikoo** [comment]
Support for `projectDir` was released just with v2.45.0

Having that you may shift project boundary via `projectDir` setting, e.g. @m-radzikowski in your case configuring `projectDir: ../` should fix the problem and deprecation will no longer be shown.

`projectDir` comes with no side effects, and will not introduce any in a future (as explained in top description it'll serve as project directory reference for other functionalities with which we may want to reach out beyond service directory)

**Comment by JohnnerJAL** [comment]
Hello everybody. Medikoo I am new to using serverless. So could you help me to solve the warning and the problem? I have tried to do the things you propose as a solution, but it does not work for me. I think I'm doing something wrong.

I have the following folder structure:

projectRoot/
  services/
    service1/
      lambda1/
        handler.js
      serverless.yml
    service2/
      lambda1/
        handler.js
      serverless.yml
  serverless.common.yml
_______________________________________________

With your proposed solution I tried on the (projectRoot/services/service1/serverless.yml) file:

service: myservice1

projectDir: ../../

custom: ${file(../../serverless.common.yml):custom}

provider: ....

functions: ....

plugins: ....

_________________________________________

Deprecation warning: Variables resolver reports following resolution errors:
              - Cannot resolve variable at "custom": Cannot load file from outside of service folder
            From a next major this will be communicated with a thrown error.
            Set "variablesResolutionMode: 20210326" in your service config, to adapt to new behavior now
            More Info: https://www.serverless.com/framework/docs/deprecations/#NEW_VARIABLES_RESOLVER


TypeError: Cannot create property 'webpack' on string '${file(../../serverless.common.yml):custom}'

Thanks so much for your time

**Comment by medikoo** [comment]
@JohnnerJAL can you open a new discussion here https://github.com/serverless/serverless/discussions ? Also please provide formatted input (so YML content is readable, when it's plain text it's hard to investigate what could be the potential issue. Additionally it'll be nice if you prepare as small as possible reproduction case

**Comment by Bolik777** [comment]
setting the projectDir variable does have side effects. I don't get " - Cannot resolve variable at" warnings anymore, but other warnings being displayed "Parameter name: can't be prefixed with "ssm" (case-insensitive). If formed as a path, it can consist of sub-paths divided by slash symbol; each sub-path can be formed as a mix of letters, numbers and the following 3 symbols .-_,". Now it doesn't like referencing SSM parameters in lambda environment.

**Comment by medikoo** [comment]
@Bolik777 messages you show do not seem related to `projectDir` setting. If you feel there's a bug, can you open a new bug request? Alternatively please use https://github.com/serverless/serverless/discussions for general Q &A

**Comment by Bolik777** [comment]
these messages appear only if i set the projectDir

**Comment by medikoo** [comment]
> these messages appear only if i set the projectDir

By setting `projectDir`, you've allowed to resolve some `file` variables with a new resolver, and in turn some other variables were probably also picked by new resolver. 

New resolver, in some places, has slightly changed rules. If from presented warning it's not clear what's the problem please open a new bug report, respecting all it's remarks (note: this is not related to this issue).

Note also that if new resolver has problems with resolution, resolution will fallback to old resolver, so ideally there should be no difference, aside of some deprecations warnings being displayed encouraging to adapt fully to new resolver.

**Comment by Bolik777** [comment]
thanks for the explanation. i am not resolving file variables, the warning comes for ssm parameter referenced in lambda environment. 
so i have:
`myFunction:
  handler: functions/function.handler
  memorySize: 128
  environment: 
    my_var: ${ssm:/path/to/var~false}
 `
and the warning is about the ssm parameter:
` - Cannot resolve variable at "functions.2.myFunction.environment.my_var": Parameter name: can't be prefixed with "ssm" (case-insensitive). If formed as a path, it can consist of sub-paths divided by slash symbol; each sub-path can be formed as a mix of letters, numbers and the following 3 symbols .-_,`

**Comment by medikoo** [comment]
@Bolik777 this is not related to this issue, so let's not discuss it here. Can you open new [discussion](https://github.com/serverless/serverless/discussions) or bug report with reproduction details

**Comment by poldark101** [comment]
Hi everyone, I'm also working with a multi-service repo.  I have upgraded to 2.45.0 and have added projectDir to my serverless.yml file.  This resolved my resources. ie. Its now referencing variables outside of the serviceDir.  However, now I am having an issue with the service name.

**serverless.yml**
```
service: ${file(../../deploy/serverless/config.${self:custom.stage}.yml):servicePrefix}-xyz
projectDir: ../../
variablesResolutionMode: 20210326
```

I get the following error on deploy
_Serverless: Configuration warning at 'service.name': should match pattern "^[a-zA-Z][0-9a-zA-Z-]+$"_
  ..
_The stack service name "${file(../../deploy/serverless/config.all.yml):servicePrefix}-xyz-dev" is not valid. A service name should only contain alphanumeric (case sensitive) and hyphens. It should start with an alphabetic character and shouldn't exceed 128 characters._

I've tried using projectDir in place of ../../ however this is not the issue.  It's not resolving the value within $file.  Can the service name be defined in this way still?

Project structure:
```
services
--service1
   --serverless.yml
--service2
   --serverless.yml
deploy
--serverless
```

Thanks

**Comment by medikoo** [comment]
@poldark101 I'm not able to reproduce that. Ensure that you're relying on latest version of a Framework, and if you still see the issue, please open a new bug report (with complete answers to all template remarks, so we have all needed information to reproduce the issue on our side)

**Comment by poldark101** [comment]
@medikoo  Also happening on 2.46.0. I will open a new bug. Thanks

**Comment by 7wQvRTU2** [comment]
oops. 

https://twitter.com/austencollins/status/1478488004121354244

**Comment by mnapoli** [comment]
Hey everyone, after discussing internally here's an update: the original deprecation ("Cannot load file from outside of service folder") [is being removed](https://github.com/serverless/serverless/pull/10478) from Serverless Framework v2.

Including from parent directories (`${file(../...`) will work in v2 and v3 by default. The `projectDir` option will no longer be required (the `projectDir` option will actually be ignored, you'll be able to remove it from your projects).

**Comment by mnapoli** [comment]
Quick update on Compose: we released [Serverless Framework Compose](https://www.serverless.com/blog/serverless-framework-compose-multi-service-deployments) last week  🎉 

Again, to avoid confusion, Compose doesn't directly solve _packaging_ shared code (using Lerna/Nx/Turborepo/etc. is a good solution), but it should help sharing outputs between services, which should simplify the configuration in some cases. So I thought I should mention it again here.

**Comment by jeanbmar** [comment]
Let's hope for https://github.com/npm/cli/pull/4082 that would help immensely to package shared code.

**Comment by ZeoKnight** [comment]
Not sure if appropriate here (apologies if not) but it might be worth Serverless adding a notice on the following guide if it's no longer advised to implement multi-service applications (with multiple nested serverless.yml files). 

https://www.serverless.com/blog/structuring-a-real-world-serverless-app
